### PR TITLE
[SR-6996]Better diagnostic when declaring a class inheritance with parentheses

### PR DESCRIPTION
**Pull request by Kazutaka333**
This PR adds a diagnosis message specifically for the case where python style inheritance clause is used and also FixIt option.
`class Foo (Int) {}`
↓↓↓↓↓
`class Foo : Int {}`

Resolves [SR-6996](https://bugs.swift.org/browse/SR-6996).

### CODE REVIEW

File: include/swift/AST/DiagnosticsParse.def [hunk 1]
```diff
@@ -362,6 +362,8 @@ ERROR(expected_lbrace_class,PointsToFirstBadToken,
       "expected '{' in class", ())
 ERROR(expected_rbrace_class,none,
       "expected '}' in class", ())
+ERROR(expected_colon_class, PointsToFirstBadToken,
+      "expected ':' before inheritance identifier",())
```
Comment (rintaro) [inline include/swift/AST/DiagnosticsParse.def]: "inheritance clause" please.

File: test/Parse/ConditionalCompilation/decl_parse_errors-3.swift [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// RUN: %target-typecheck-verify-swift                                           
+
+// expected-error @+1 {{expected ':' before inheritance identifier}} {{10-11=: }} {{14-15=}}
+class C3 (Int) { 
+}
```
Comment (rintaro) [inline test/Parse/ConditionalCompilation/decl_parse_errors-3.swift]: In addition to @CodaFi's comment, could you add more tests? such as:
```
class C1(Base<Int>) {}
class C2<T>(Foo.Bar) where T:AnyObject {}
```

File: include/swift/AST/DiagnosticsParse.def (hunk 1 above)
Comment (CodaFi) [inline include/swift/AST/DiagnosticsParse.def]: Perhaps `expected ‘:’ to delimit inheritance clause`?

File: test/Parse/ConditionalCompilation/decl_parse_errors-3.swift [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// RUN: %target-typecheck-verify-swift                                           
+
+// expected-error @+1 {{expected ':' before inheritance identifier}} {{10-11=: }} {{14-15=}}
+class C3 (Int) {
```
Comment (CodaFi) [inline test/Parse/ConditionalCompilation/decl_parse_errors-3.swift]: Please move this test somewhere more relevant.  Maybe `Parse/recovery.swift`

**Comment by CodaFi** [comment]
Just a few nits, nothing more.  Good work!

**Review by rintaro — COMMENTED** [review]
Pretty good start, Thanks!

**Comment by Kazutaka333** [comment]
@CodaFi 
@rintaro 
I made the change as you have suggested and added more tests. Could you review it when you get a chance?

File: lib/Parse/ParseDecl.cpp [hunk 4]
```diff
@@ -5666,13 +5666,24 @@ ParserResult<ClassDecl> Parser::parseDeclClass(SourceLoc ClassLoc,
...
+        Status.setIsParseError();
+        return Status;
+      }
+      SourceLoc RParenLoc;
+      consumeIf(tok::r_paren, RParenLoc);
       diagnose(LParenLoc, diag::expected_colon_class)
-        .fixItReplace(LParenLoc, ": ")
-        .fixItRemove(RParenLoc);
+      .fixItReplace(LParenLoc, ": ")
+      .fixItRemove(RParenLoc);
```
Comment (rintaro) [inline lib/Parse/ParseDecl.cpp]: Nitpick: indent.

File: test/Parse/recovery.swift [hunk 5]
```diff
@@ -555,6 +555,31 @@ class WrongDeclIntroducerKeyword1 { // expected-note{{in declaration of 'WrongDe
...
+class WrongInheritanceClause2(Base2<Int>) {}
+
+// expected-error @+1 {{expected ':' to begin inheritance clause}} {{33-34=: }} {{49-50=}}
+class WrongInheritanceClause3<T>(SubModule.Base1) where T:AnyObject {}
+
+// expected-error @+1 {{expected ':' to begin inheritance clause}} {{30-31=: }} {{51-52=}}
+class WrongInheritanceClause4(SubModule.Base2<Int>) {}
+
+// expected-error @+1 {{expected ':' to begin inheritance clause}} {{33-34=: }} {{54-55=}}
+class WrongInheritanceClause5<T>(SubModule.Base2<Int>) where T:AnyObject {}
```
Comment (rintaro) [inline test/Parse/recovery.swift]: Could you add a few missing-`)` test case?
```swift
class Foo(Bar {}
class Foo<T>(Bar where T:AnyObject {}
```

**Review by rintaro — COMMENTED** [review]
Looks great! A few nitpicks.

**Comment by Kazutaka333** [comment]
@rintaro I added tests for missing parenthesis. thank you for the review!

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by rintaro — APPROVED** [review]
Thanks!

**Pull request merged**